### ISSUE TITLE
Cart Price Rule based on weight does not work

### ISSUE THREAD

**Issue by ghost** [magento/magento2#6584]
When making a cart price rule using weight for the condition, the rule does not apply in the cart correctly. All the products have weights and flat rate shipping option is set to $10. Weight is set to kg.
### Preconditions

1. Magento CE 2.1.1
2. CentOS 6
3. PHP 5.6.25
4. MySQL 5.6
### Steps to reproduce

1. Create new cart price rule for all customer groups
2. Set condition as the following:
   
   If ALL  of these conditions are TRUE :
               Total Weight  equals or less than  10
3. Leave actions settings defualt except Free Shipping. Set Free Shipping to For shipment with matching items
4. Save rule
### Expected result

1. Shopping cart should show free shipping when product weight is under 10
### Actual result

1. Shopping cart always shows free shipping

**Comment by sevos1984** [comment]
Thanks for reporting, internal ticket created MAGETWO-58334

**Comment by licentia** [comment]
File: magento2/app/code/Magento/Quote/Model/Quote/Address/Total/Shipping.php:57

You guys are setting the weight to zero before checking the rule.

```php
$address->setWeight(0);
$address->setFreeMethodWeight(0);
$addressWeight = $address->getWeight();
$freeMethodWeight = $address->getFreeMethodWeight();
$address->setFreeShipping(
    $this->freeShipping->isFreeShipping($quote, $shippingAssignment->getItems())
);
```

 Move the lines:

```php
$address->setWeight(0);
$address->setFreeMethodWeight(0);
$addressWeight = $address->getWeight();
$freeMethodWeight = $address->getFreeMethodWeight();
```

below:

```php
$address->setFreeShipping(
    $this->freeShipping->isFreeShipping($quote, $shippingAssignment->getItems())
);
```

**Comment by SerhiyShkolyarenko** [comment]
MAGETWO-58334 is delivered to develop branch. Closing issue.

**Comment by Ctucker9233** [comment]
@SerhiyShkolyarenko Will MAGETWO-58334 be included in 2.1.3?

**Comment by tareefj** [comment]
We're still having issues with this case:

**Setting Cart Price Rule based on weight does not work**

_Preconditions_
Magento CE 2.1.3

_Steps to reproduce_

1. Create new cart price rule for all customer groups

2. Set condition as the following:

If ALL of these conditions are TRUE :
Total Weight  less than  2.5  
Subtotal  equals or greater than  499  

3. Leave actions settings defualt except Free Shipping. Set Free Shipping to For shipment with matching items

4. Save rule

_Expected result_

Shopping cart should show free shipping when product weight is under 2.5 kg and over 499

_Actual result_

Shopping cart always shows free shipping when price is greater than 499 (even if product weight is greater than 2.5 kg)
